app/vmalert/notifier/template_func.go

```diff
@@ -14,21 +14,40 @@
 package notifier
 
 import (
+	"errors"
 	"fmt"
-	html_template "html/template"
 	"math"
 	"net/url"
 	"regexp"
 	"strings"
-	text_template "text/template"
 	"time"
+
+	htmlTpl "html/template"
+	textTpl "text/template"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/app/vmalert/datasource"
 )
 
-var tmplFunc text_template.FuncMap
+// QueryFn is used to wrap a call to datasource into simple-to-use function
+// for templating functions.
+type QueryFn func(query string) ([]datasource.Metric, error)
+
+func funcsWithQuery(query QueryFn) textTpl.FuncMap {
```

Comment:
as far as i understand, it will make copy of template funcs each time template is executed. Maybe its better to store `textTpl.FuncMap` per `executor` struct?

Reply:
Yes, it will. It may worth to apply optimisation that you proposed.